test/e2e/node_reuse_test.go

```diff
@@ -0,0 +1,525 @@
+			bmhs := bmo.BareMetalHostList{}
+			Expect(targetClusterClient.List(ctx, &bmhs, client.InNamespace(namespace))).To(Succeed())
+			bmhConfMap := map[string]bmo.BareMetalHost{}
+			for _, v := range bmhs.Items {
+				bmhConfMap[v.Name] = v
+			}
+			for _, item := range deprovisioning_bmh {
+				if bmhConfMap[item.Name].Status.Provisioning.State != bmo.StateReady {
+					return errors.New("A bmh is not ready yet")
+				}
+			}
+			return nil
+		},
+		e2eConfig.GetIntervals(specName, "wait-bmh-deprovisioning-ready")...,
+	).Should(Succeed())
+
+	By("Unmarking all the ready BMHs with unhealthy annotation")
+	bmhs = bmo.BareMetalHostList{}
+	Expect(targetClusterClient.List(ctx, &bmhs, client.InNamespace(namespace))).To(Succeed())
+	for _, item := range bmhs.Items {
+		if item.Status.Provisioning.State == bmo.StateReady {
+			annotateBmh(ctx, targetClusterClient, item, "capi.metal3.io/unhealthy", nil)
+		}
+	}
+
+	By("Checking if just deprovisioned became ready BMH and re-used for next provisioning")
+	Eventually(
+		func() error {
+			bmhs := bmo.BareMetalHostList{}
+			Expect(targetClusterClient.List(ctx, &bmhs, client.InNamespace(namespace))).To(Succeed())
+			bmhConfMap := map[string]bmo.BareMetalHost{}
+			for _, v := range bmhs.Items {
+				bmhConfMap[v.Name] = v
+			}
+			for _, item := range deprovisioning_bmh {
+				if bmhConfMap[item.Name].Status.Provisioning.State != bmo.StateProvisioning {
+					return errors.New("A bmh is not provisioning yet")
+				}
+			}
+			return nil
+		},
+		e2eConfig.GetIntervals(specName, "wait-bmh-ready-provisioning")...,
+	).Should(Succeed())
+
+	Byf("Waiting until worker machine becomes running and updated with new %s k8s version", upgradedK8sVersion)
+	Eventually(
+		func() int {
+			machines := &clusterv1.MachineList{}
+			Expect(targetClusterClient.List(ctx, machines, client.InNamespace(namespace))).To(Succeed())
+			running_upgraded_len := 0
+			for _, machine := range machines.Items {
+				if machine.Status.GetTypedPhase() == clusterv1.MachinePhaseRunning && *machine.Spec.Version == upgradedK8sVersion {
+					running_upgraded_len++
+					Logf("Machine [%v] is upgraded to (%v) and running", machine.Name, upgradedK8sVersion)
+				}
+			}
+			return running_upgraded_len
+		},
+		e2eConfig.GetIntervals(specName, "wait-machine-running")...,
+	).Should(Equal(2))
+
+	By("Getting provisioned BMH names and UUIDs after upgrade in MachineDeployment")
+	mdBmhAfterUpgrade := getProvisionedBmhNamesUuids(targetClusterClient)
+
+	By("Checking difference between before and after upgrade mappings in MachineDeployment")
+	equal = reflect.DeepEqual(mdBmhBeforeUpgrade, mdBmhAfterUpgrade)
+	Expect(equal).To(BeTrue(), "The same BMHs were not reused in MachineDeployment")
+
+	By("Scaling controlplane up to 3")
+	scaleControlPlane(ctx, targetClusterClient, client.ObjectKey{Namespace: namespace, Name: clusterName}, 3)
+
+	Byf("Waiting until all %d bmhs are provisioned", numberOfAllBmh)
+	Eventually(func() error {
```

Comment:
Please make all `Eventually` follow the same style. The style of line 409 is preferred since it can save some lines.